modules/core/04-channel/types/upgrade.go

```diff
@@ -69,3 +71,31 @@ func (uf UpgradeFields) ValidateBasic() error {
 func (ut UpgradeTimeout) IsValid() bool {
 	return !ut.Height.IsZero() || ut.Timestamp != 0
 }
+
+// HasPassed returns true if the upgrade has passed the timeout height or timestamp.
+// A string is returned containing information about the timeout relative to the context's time and height.
+func (ut UpgradeTimeout) HasPassed(ctx sdk.Context) (bool, string) {
```

Comment:
Is this the cleanest return signature? feels like an odd pattern. Why don't we just return an sdk error with an error string

Reply:
because there was [a preference for an info string](https://github.com/cosmos/ibc-go/pull/3535#discussion_r1183460674) rather than returning an error if the timeout had passed

Reply:
do you mean an `error` only? no bool? @AdityaSripal